In [21]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

df = pd.read_csv("cl_flipkart_data.csv")
df.head()

unnamed:_0                                   product_name  \
0           0  Samsung Galaxy A35 5G (Awesome Lilac, 128 GB)   
1           1                                      Jio F320B   
2           2   Samsung Galaxy A35 5G (Awesome Navy, 128 GB)   
3           3                        Ai+ Pulse (Blue, 64 GB)   
4           4                    Ai+ Nova 5G (Black, 128 GB)   

   discount_percentage                         camera   battery  \
0                 44.0  50mp + 8mp + 5mp | 13mp front  5000 mah   
1                 17.0                     0.5mp rear  2000 mah   
2                 44.0  50mp + 8mp + 5mp | 13mp front  5000 mah   
3                 18.0         50mp rear  | 5mp front  5000 mah   
4                 22.0         50mp rear  | 5mp front  5000 mah   

                        display  ratting_number  selling_price actual_price  \
0  16.76 cm (6.6 inch) full hd+             4.4          18999     33,99944   
1             6.1 cm (2.4 inch)             3.7           2472      2,99917   
2  16.76 cm (6.6 inch) full hd+             4.4          18999     33,99944   
3     17.13 cm (6.745 inch) hd+             4.3           6499      7,99918   
4     17.13 cm (6.745 inch) hd+             4.1           8499     10,99922   

  rattings  reviews Price_Band Expandable  RAM    ROM  
0   67,499   3897.0  Mid-Range        1tb  8gb  128gb  
1    5,825    337.0     Budget        0gb  0gb    4gb  
2   67,499   3897.0  Mid-Range        1tb  8gb  128gb  
3   40,717   2496.0     Budget        1tb  4gb   64gb  
4   19,729   1327.0     Budget        1tb  6gb  128gb

In [43]:
X = df[['rattings','reviews']]
y = df['actual_price']

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

ValueError: could not convert string to float: '19,729' 
so convert string to float

In [44]:
for col in X_train.columns:
    if X_train[col].dtype == 'object':
        X_train[col] = (
            X_train[col]
            .astype(str)
            .str.replace(',', '', regex=True)
        )


In [45]:
X_train = X_train.apply(pd.to_numeric, errors='coerce')
X_test  = X_test.apply(pd.to_numeric, errors='coerce')

In [46]:
X_train = X_train.fillna(0)
X_test = X_test.fillna(0)


ValueError: Input contains NaN
Data contains Nans so i remove NaNs value....

In [48]:
print("X_train NaNs:\n", X_train.isna().sum())
print("\ny_train NaNs:\n", y_train.isna().sum())


X_train NaNs:
 rattings    0
reviews     0
dtype: int64

y_train NaNs:
 45


In [50]:
y_train = (
    y_train.astype(str)
           .str.replace(',', '', regex=True)
           .str.extract(r'(\d+)')[0]
           .astype(float)
)


In [51]:
y_train = y_train.fillna(y_train.median())


In [52]:
print("X NaNs:", X_train.isna().sum().sum())
print("y NaNs:", y_train.isna().sum())
print("y dtype:", y_train.dtype)


X NaNs: 0
y NaNs: 0
y dtype: float64


In [53]:
model = LinearRegression()
model.fit(X_train, y_train)

LinearRegression()

In [54]:
y_test = (
    y_test.astype(str)
          .str.replace(',', '', regex=True)
          .str.extract(r'(\d+)')[0]
          .astype(float)
)


In [55]:
y_test = y_test.fillna(y_train.median())


In [56]:
print(y_test.dtype)
print("NaNs in y_test:", y_test.isna().sum())


float64
NaNs in y_test: 0


In [57]:
import numpy as np

y_pred = model.predict(X_test)

mse = mean_squared_error(y_test, y_pred)
rmse = np.sqrt(mse)
r2 = r2_score(y_test, y_pred)

rmse, r2

(np.float64(1845999.1997117705), 0.03932291420005851)